cp: move copy_directory() to its own module

jfinkels opened the pull request:
This pull request refactors `copy_directory()` and related helper functions into its own new module `copydir.rs`. This commit also adds some additional structs and helper functions to make the code easier to read and maintain. This commit does not change the behavior of the `copy_directory()` function, only the organization of the code.

I am proposing this change because I felt while working on pull request #3894 that it was harder than necessary for me to read and understand the code.

tertsdiepraam commented on src/uu/cp/src/copydir.rs:
```diff
@@ -0,0 +1,308 @@
+                    symlinked_files,
+                    preserve_hard_links,
+                    &mut hard_links,
+                )?;
```
Is it okay to bubble the errors up here? Shouldn't they go to the `show_error!` on lin 303?

jfinkels replied:
I was matching the previous behavior. See the main loop in the old code, starting with:
```

    for path in WalkDir::new(root)
        .same_file_system(options.one_file_system)
        .follow_links(options.dereference)
    {
```
There's a bunch of `?` and `return Err(...)` that will terminate the loop immediately and exit the `copy_directory()` function with an error. I didn't check whether this is the right behavior when compared with GNU cp, but it seemed to be what the old code was doing. The `show_error!()` on line 303 corresponds to the `or_continue!()` macro in the old code.

tertsdiepraam commented:
Looks good! I just have the one comment above but otherwise ready to merge. Could you separate the moving around and refactoring next time, so it's easier to review?

jfinkels commented:
> Could you separate the moving around and refactoring next time, so it's easier to review?

Yes, absolutely.

jfinkels commented:
I rebased and updated this branch so that one commit moves the `copy_directory()` function as-is to a new module, and the next commit breaks up the `copy_directory()` function into smaller helper functions.

sylvestre commented:
Conflicting again after the merge of your other PR ;)
